[cssom] Use shortest possible serialization for border-spacing

AlexsanderDamaceno opened the pull request:
#### 6b740c4e45079dbfba57ff6e8c558d91e490e0e7
<pre>
[cssom] Use shortest possible serialization for border-spacing
<a href="https://bugs.webkit.org/show_bug.cgi?id=284724">https://bugs.webkit.org/show_bug.cgi?id=284724</a>

Reviewed by Tim Nguyen.

Fix missing case in `border-spacing` serialisation. When the we have `border-spacing: 0px 0px`, we should apply the shortest serialization principle.
The WPT test was failing because we were serializing this configuration as &quot;0px 0px&quot; instead of &quot;0px&quot;.

* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/parsing/border-spacing-computed-expected.txt:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/inheritance-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-with-zoom-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt:
* LayoutTests/fast/css/getComputedStyle/getComputedStyle-border-spacing-expected.txt:
* LayoutTests/fast/css/table-border-spacing-expected.txt:
* LayoutTests/fast/css/table-border-spacing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-substitution-basic.html:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-attribute.html:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-types.js:
(const.lengthPairType.testAdditionOrAccumulation):
(const.lengthPairType.testInterpolation):
(const.lengthPairType.testAdditionOrAccumulation):
(const.positionType.testInterpolation):
(const.positionType.testAdditionOrAccumulation):

Canonical link: <a href="https://commits.webkit.org/289431@main">https://commits.webkit.org/289431@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f18917bbe945f10f2cac6204953284afcde9347

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86887 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6394 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41229 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91735 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37620 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88936 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6662 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14458 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67166 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24935 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89890 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5073 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78639 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47484 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4858 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33004 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36737 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75360 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33883 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93628 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14040 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10195 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75966 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14241 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74486 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75163 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18494 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19482 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17898 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/6874 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14063 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/19395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13801 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17246 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15586 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3574,6 +3574,8 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
+        if (style.horizontalBorderSpacing() == 0.0 && style.verticalBorderSpacing() == 0.0)
+            return zoomAdjustedPixelValue(style.horizontalBorderSpacing(), style);
         return CSSValueList::createSpaceSeparated(zoomAdjustedPixelValue(style.horizontalBorderSpacing(), style),
             zoomAdjustedPixelValue(style.verticalBorderSpacing(), style));
```
The lengths should coalesce in all cases where the horizontal value is equal to the vertical value.

`CSSValuePair::create(zoomAdjustedPixelValue(style.horizontalBorderSpacing(), style), zoomAdjustedPixelValue(style.verticalBorderSpacing(), style));` gives you this functionality for free.

nt1m replied:
(e.g. `1px 1px` should collapse to `1px` as well)

AlexsanderDamaceno replied:
I think there’s some confusion here with the CSSValuePair. In this case, the code uses CSSValueList, and I don’t think there’s a function that compares the values in this class, is right?.

nt1m replied:
You can just replace all those lines to:
```
return CSSValuePair::create(zoomAdjustedPixelValue(style.horizontalBorderSpacing(), style), zoomAdjustedPixelValue(style.verticalBorderSpacing(), style));
```

The serialization of `CSSValuePair` already coalesces identical values (unless you use `CSSValuePair::createNonCoalescing`), see: https://searchfox.org/wubkat/rev/3c50bf12bcc7cc7cb47c8b3a8d238e5f433dcf35/Source/WebCore/css/CSSValuePair.cpp#64-71

AlexsanderDamaceno replied:
Ok great,  Thank you.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-tables/parsing/border-spacing-computed-expected.txt:
```diff
@@ -1,6 +1,6 @@
 
 PASS Property border-spacing value '10px 20px'
-PASS Property border-spacing value '0'
+FAIL Property border-spacing value '0' assert_equals: expected "0px 0px" but got "0px"
```
We should probably re-import the WPT before doing this, it's kinda strange that this appears as a regression

AlexsanderDamaceno replied:
@nt1m  It is strange that the expectation in WebKit repository from this test is to passing all tests here, but when we try to execute the border-spacing-computed.html on the WPT site, it shows that it is failing.

http://wpt.live/css/css-tables/parsing/border-spacing-computed.html

AlexsanderDamaceno replied:
This WPT was imported last week, so I think is updated.

nt1m replied:
Oh it's marked as failing here: https://searchfox.org/wubkat/rev/3c50bf12bcc7cc7cb47c8b3a8d238e5f433dcf35/LayoutTests/TestExpectations#6641 , you should remove that expectation in this patch now that you've made it pass

nt1m replied:
You've also progressed the test above, so you can remove that as well: https://searchfox.org/wubkat/rev/3c50bf12bcc7cc7cb47c8b3a8d238e5f433dcf35/LayoutTests/TestExpectations#6640

nt1m reviewed (approved):
Can you adjust the commit title to match the Bugzilla title?

nt1m reviewed (approved):
Looks good! The title is still not fully accurate though, something like this would make more sense:

"[cssom] Use shortest possible serialization for border-spacing"

Reason being that the border-spacing serialization is still space-separated when there are 2 values.

Ahmad-S792 commented on LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-types.js:
```diff
@@ -200,7 +200,7 @@ const lengthPairType = {
         animation,
         idlName,
-        [{ time: 0, expected: '20px 20px' }]
+        [{ time: 0, expected: '20px' }]
```
It might need more changes like - https://github.com/web-platform-tests/wpt/commit/b9f5f525800cf1cfbdb46acd2d1bd1d72feb3d11#diff-be868bff8396deaf6494fa52c78c79bb327ac91f88de608413906d48c258b45d

Ahmad-S792 replied:
^ This would fix your `web-animations` tests.

nt1m replied:
@AlexsanderDamaceno @Ahmad-S792's suggestion would fix your failure, essentially perspective-origin (a different CSS property) ends up also expecting 1 instead of two values, what Ahmad suggested would allow special casing border-spacing.

AlexsanderDamaceno replied:
@Ahmad-S792  Thank you for the sugestion.

nt1m commented on LayoutTests/fast/css/getComputedStyle/getComputedStyle-border-spacing-expected.txt:
```diff
@@ -2,16 +2,16 @@ Test calling getPropertyValue on computed styles for -webkit-backround-size prop
 
 getPropertyValue(border-spacing): 10px 20px
 getPropertyCSSValue(border-spacing): [object CSSValueList]
```
```suggestion
getPropertyCSSValue(border-spacing): [object CSSPrimitiveValue]
```

nt1m commented on LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt:
```diff
@@ -196,7 +196,7 @@ transition-delay: 0s
 transition-property: all
 transition-timing-function: ease
-unicode-bidi: isolate
+unicode-bidi: normal
```
This is unrelated

AlexsanderDamaceno committed:
[cssom] Use shortest possible serialization for border-spacing
https://bugs.webkit.org/show_bug.cgi?id=284724

Reviewed by Tim Nguyen.

Fix missing case in `border-spacing` serialisation. When the we have `border-spacing: 0px 0px`, we should apply the shortest serialization principle.
The WPT test was failing because we were serializing this configuration as "0px 0px" instead of "0px".

* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/parsing/border-spacing-computed-expected.txt:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-tables/inheritance-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-with-zoom-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-without-renderer-expected.txt:
* LayoutTests/fast/css/getComputedStyle/getComputedStyle-border-spacing-expected.txt:
* LayoutTests/fast/css/table-border-spacing-expected.txt:
* LayoutTests/fast/css/table-border-spacing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-variables/variable-substitution-basic.html:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-attribute.html:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-types.js:
(const.lengthPairType.testAdditionOrAccumulation):
(const.lengthPairType.testInterpolation):
(const.lengthPairType.testAdditionOrAccumulation):
(const.positionType.testInterpolation):
(const.positionType.testAdditionOrAccumulation):

Canonical link: https://commits.webkit.org/289431@main
